tests/utils.py

```diff
@@ -66,3 +68,77 @@ def deco(f):
 
         return awaiter
     return deco
+
+
+class MockConnectionContext:
+    def __init__(self, db):
+        self._db = db
+
+    async def __aenter__(self):
+        await self._db._lock.acquire()
+        return self._db._connection
+
+    async def __aexit__(self, exc_type, exc, tb):
+        self._db._lock.release()
+
+
+class MockDatabase:
+    """
+    This class mocks the FAFDatabase class, rolling back all transactions
+    performed during tests. To do that, it proxies the real db engine, giving
+    access to a single connection the results of which are never comitted.
+    Since the server uses that single connection, it sees all changes made, but
+    at the same time we can rollback all these changes once the test is over.
+
+    Note that right now the server relies on autocommit behaviour of aiomysql.
+    Any future manual commit() calls should be mocked here as well.
+    """
+    def __init__(self, loop):
+        self._loop = loop
+        self.engine = None
+        self._connection = None
+        self._conn_present = Event()
+        self._keep = None
+        self._lock = Lock()
+        self._done = Event()
+
+    async def connect(self, host='localhost', port=3306, user='root',
+                      password='', db='faf_test', minsize=1, maxsize=1,
+                      echo=True):
+        if self.engine is not None:
+            raise ValueError("DB is already connected!")
+        self.engine = await create_engine(
+            host=host,
+            port=port,
+            user=user,
+            password=password,
+            db=db,
+            autocommit=False,
+            loop=self._loop,
+            minsize=minsize,
+            maxsize=maxsize,
+            echo=echo
+        )
+        self._keep = self._loop.create_task(self._keep_connection())
+        await self._conn_present.wait()
+
+    async def _keep_connection(self):
+        async with self.engine.acquire() as conn:
+            self._connection = conn
+            self._conn_present.set()
+            await self._done.wait()
+            self._connection = None
+
+    def acquire(self):
+        return MockConnectionContext(self)
+
+    async def close(self):
+        if self.engine is None:
+            return
+
+        async with self._lock:
+            self._done.set()
+            await self._keep
+            self.engine.close()
+            await self.engine.wait_closed()
+            self.engine = None
```

Comment:
Did you know, that you don't need that class at all?
See here, how to work with database as yield fixture:
https://github.com/norraxx/faf-aio-replayserver/blob/master/tests/fixtures/db_fixtures.py#L7
And if you'll need cursor, you can use that one fixture directly.
https://github.com/norraxx/faf-aio-replayserver/blob/master/tests/fixtures/db_fixtures.py#L48

You don't need to "overwrite" database connector behavior, let it be. Just ensure call `begin` at the beginning of the test, and `rollback` at the end.

That will allow you access to the database, make queries, etc.. during whole test, and it will clean everything else at the end of the each test.

Reply:
This gives me a single connection. The server expects to have access to an entire pool. The class is intended to mock that pool with a single connection.

Reply:
Check what I've send again and think how it works and then think what you've wrote.

And then you'll understand that if you'll use my way, you won't need to touch `async with self._db.engine.acquire() as conn` and your tests will work properly.

Anyway: I'm impressed how you understand code, very fast, in 9 minutes you did great analysis to approve your solution to not to change and do anything else. ;-)

I think, I'll make Mazor's reject counter, and add there time per rejection.

Reply:
Your solution creates a connection with a different interface - not the sqlalchemy-wrapped connection that the lobby server uses, but one used to directly execute SQL strings. The only way I found in the API to create the former is to create an engine with a pool, then grab a connection from that pool. If you know of a better way to create and provide an `SAConnection` instance, please suggest one.

Furthermore, your solution:
* Utilizes module-level globals for testing, which risks inter-test interaction,
* Has separate code paths in non-testing code for running tests, which is a horrible practice.

I don't intend to do either of these things.

I'd also like you to use an impersonal tone during reviews, as I do - the purpose of these is to be informative and direct so code can be improved. I have no time and no interest in guessing what I'm "supposed to understand" that you're withholding from me, nor do I care about any clever remarks unrelated to code itself.